### DISCUSSION TITLE
For loops, Paths and plots

### DISCUSSION THREAD

**Discussion by maufl** [slint-ui/slint#776] [Q&A]
Hey, may thanks for this great library! I like really like how it's looking so far and how easy it is to have a GUI in a Rust program now.

I saw the IoT dashboard example and would like to build a similar looking dashboard for myself (showing mostly the balance in my accounts).
My initial idea was to use `Path` and a `for` expression to build a plot line from an array of data points. I quickly realized that it is probably not possible the way I want to do it.
1. To calculate the control points I would want to have access to sibling points but that is afaik not possible.
2. I might want to handle the first and last point differently to paint a area instead of just a line. I can compare against the index but I can't pick a different component (e.g. `MoveTo` instead of `QuadraticTo`) afaik.

Then I tried to use the `plotters` crate. I was quickly able to get an example working, but realized that SixtyFPS only takes pixel images, not vector graphics as `image`. My plots would therefore not scale with the window.

My questions are:
1. Is the `for` expression limited like this for a reason (i.e. compile time optimization) or are there plans to make it more general like in Rust?
2. Will it be possible to render vector graphics?

Thanks again for your work! I think this might have a huge impact on the adoption of Rust :)

**Comment by Vollbrecht** [comment]
i think i am working on something real similar what you trying to archive :D it is absolutely possible but there are some quarks i first needed to learn to get where i am now. The code is currently a jumbled mess because i am still in the stage to verify if my idea is possible but here is a sneak peak of the output.

https://user-images.githubusercontent.com/12041081/147791136-c3a80db7-bc29-4a9f-8641-165f3ad9ade4.mp4

The next thing i need to create are some cursors where i can drag and drop on the bezier curve control points. And hit detection is still not 100% flashed out ;D

Did you already had a go with the memory game example ? There you find some good hints how to approach part of my solution.

**Reply by tronical** [reply]
I think it is desirable to support `for` with `Path` in the future and I don’t see a fundamental problem with it - it is just not implemented at the moment.

The `Path` element has an internal property of path elements, which are currently determined at compile time but the implementation allows for it to be dynamic. 

So for the commands string property we can offer a dynamic version that maps the output of the string binding to the internal representation of elements by calling a runtime function.

For the path sub-elements like `LineTo` etc this could be provided by a model and mapped similarly.

It’s a bit of work, but I think it’s doable.

**Comment by Be-ing** [comment]
Iterating with `for` inside `Path` currently does not work: #754. For now I am hacking around that by rendering into a `sixtyfps::Image` with tiny_skia.

**Reply by ogoffart** [reply]
`Path{ for ... }` does not work yet,  but  `for xxx in model : Path { ... } ` works. 
I know that this is quite limited, but can already be of some use.
